pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+
+        debug!(
+            target: "runtime",
+            "Epoch {} begins at slot {}",
+            current_epoch_no, current_slot_no
+        );
+        Epoch::put(current_epoch_no);
+        Epochs::insert(
+            current_epoch_no,
+            (active_validator_count, current_slot_no, None as Option<u64>),
+        );
+    }
+}
+
+/// Indicates to the session module if the session should be rotated.
+impl<T: Trait> pallet_session::ShouldEndSession<T::BlockNumber> for Module<T> {
+    fn should_end_session(_now: T::BlockNumber) -> bool {
+        print("Called should_end_session");
+
+        // TODO: Next 2 are debugging lines. Remove them.
+        let current_block_no = <system::Module<T>>::block_number().saturated_into::<u32>();
+        debug!(target: "runtime", "current_block_no {}", current_block_no);
+
+        let current_slot_no = match Self::current_slot_no() {
+            Some(s) => s,
+            None => {
+                print("Cannot fetch slot number");
+                return false;
+            }
+        };
+
+        let epoch_ends_at = Self::epoch_ends_at().saturated_into::<u64>();
+        debug!(
+            target: "runtime",
+            "epoch ends at {}",
+            epoch_ends_at
+        );
+
+        // Unless the epoch has had the required number of blocks, or hot swap is triggered, continue the session.
+        let hot_swap = <HotSwap<T>>::take();
+        if (current_slot_no > epoch_ends_at) || hot_swap.is_some() {
+            let (active_validator_set_changed, active_validator_count) =
+                match Self::swap_if_needed(hot_swap) {
+                    Some(t) => (true, t),
+                    None => Self::update_active_validators_if_needed(),
+                };
+
+            if active_validator_set_changed {
+                // Manually calling `rotate_session` will make the new validator set change take effect
+                // on next session (as `rotate_session` will be called again once this function returns true)
+                <pallet_session::Module<T>>::rotate_session();
+            }
+
+            let last_slot = Self::set_current_epoch_end(current_slot_no, active_validator_count);
+            debug!(
+                target: "runtime",
+                "epoch will ends at {}",
+                last_slot
+            );
+            true
+        } else {
+            false
+        }
+    }
+}
+
+/// Provides the new set of validators to the session module when session is being rotated.
+impl<T: Trait> pallet_session::SessionManager<T::AccountId> for Module<T> {
+    // SessionIndex is u32 but comes from sp_staking pallet. Since staking is not needed for now, not
+    // importing the pallet.
+
+    fn new_session(session_idx: u32) -> Option<Vec<T::AccountId>> {
+        print("Called new_session");
+        let validators = Self::active_validators();
+        // Check for error on empty validator set. On returning None, it loads validator set from genesis
```

Comment:
I'm looking at https://github.com/paritytech/substrate/blob/e824e8ab0fadec9949ebb8b9e14d98703d6b8d44/frame/session/src/lib.rs#L186 and my interpretation is that returning None just preserves the current validator set. Why do you think it loads the genesis validator set?

Reply:
You are right. Wrong comment.

Reply:
tags: documentation

Reply:
Updated the function and comment in [another branch](https://github.com/docknetwork/dock-substrate/blob/emission/pallets/poa/src/lib.rs#L1230).